### ISSUE TITLE
Permission denied when trying to access kerberos-protected file system

### ISSUE THREAD

**Issue by nifuki** [jupyter/notebook#5696]
I'm trying to give my JupyterHub users access to a shared Kerberos-protected storage and I'm stuck with the following problem:

After starting a jupyter notebook a user needs to generate a credential cache via `kinit` command which should give him write access to the storage, but the jupyter server still gives these messages on any attempt of file modification through a web-file manager: 
```
[I 13:37:34.777 NotebookApp] Creating new file in /eos/user/b/FILTERED_USERNAME
[D 13:37:34.778 NotebookApp] Saving /root/eos/user/b/FILTERED_USERNAME/untitled3.txt
[W 13:37:34.779 NotebookApp] 403 POST /api/contents/eos/user/b/FILTERED_USERNAME (FILTERED_IP): Permission denied: eos/user/b/FILTERED_USERNAME/untitled3.txt
[W 13:37:34.780 NotebookApp] Permission denied: eos/user/b/FILTERED_USERNAME/untitled3.txt
[W 13:37:34.780 NotebookApp] 403 POST /api/contents/eos/user/b/FILTERED_USERNAME (FILTERED_IP) 5.61ms referer=http://FILTERED_IP:8888/tree/eos/user/b/FILTERED_USERNAME
```

The odd thing is the web-terminal gets right permissions instantly after `kinit` (as well as commands executed in the notebook cell through `!command`), but the file navigator gets access only after leaving the server idle for some time (minutes). 

A couple more experiments make me think that jupyter caches something kerberos-related on file access and refreshes it periodically:
1. I do `kinit` prior to starting the server right after server startup I have access to the storage.
2. Then I destroy credential cache with `kdestroy` and generate the new one with `kinit` -  access remains, I can still create new files, even though a new credential cache was generated.
3. Then I do `kdestroy`, try to access some file through web-navigator, get "Permission denied", do `kinit` again - access lost.

I'm completely lost and need someone to point me in the right direction to solve this problem.

Thanks

**Comment by nifuki** [comment]
I've tracked the [line ](https://github.com/jupyter/notebook/blob/master/notebook/services/contents/fileio.py#L159) that fails (with atomic writing disabled), but I still have no clue why

**Comment by kevin-bates** [comment]
In looking around I found [your post on discourse](https://discourse.jupyter.org/t/how-to-make-file-manager-use-kerberos-ticket/5712) - including here for completeness.

Jupyter Notebook, and the contents service in general,  is certainly unaware of things like Kerberos and even the underlying storage mechanism, including NFS and its variants.  These types of low-level subsystems do not typically require application-level changes.  Sure, a Kerberos ticket may need to be refreshed for a given user from time to time, but that's beyond the scope of the current notebook server.

If you figure out what additional support is required to get this working, I think it might be useful to [`jupyter_server`](https://github.com/jupyter/jupyter_server) given it is viewed as the next notebook server.  In the meantime, I'm going to close this issue but invite any updates you may have.  We can then reopen it as needed and/or transfer to `jupyter_server`.

**Comment by nifuki** [comment]
I agree, but the jupyter is the only application that is misbehaving with this FS (so far) and I'm trying to investigate from both ends: the FS and Jupyter. The FS in question is [EOS](https://github.com/cern-eos/eos) and I couldn't find any useful information in its logs, system administrators blamed the application, which is Jupyter, so I switched to digging around in Jupyter. I haven't yet contacted EOS developers, so it's probably going to be my next stop.

**Comment by kevin-bates** [comment]
I see, thanks for the update.  Yeah, it would be good to know what might be necessary as there might be ways to plumb the appropriate config options to accommodate that, and hopefully, the EOS devs can add some insights.  I don't mean to imply there's nothing jupyter _can do_.  Just that its currently something jupyter _doesn't do_.

Have you tried approaching this from the Hub side of things since that's what creates the notebook server instance in the first place?  Perhaps there's some form of initialization that can be configured as part of spawning each server?

**Comment by nifuki** [comment]
I was able to reproduce the same behavior with a simple python script that just runs io.open in a loop, so it's definitely on the FS side and out of scope of Jupyter. My best guess is that EOS checks the credentials only once while establishing the connection and then reuses this connection with the first credential check result no matter what it was. Anyway, thanks for the discussion and I'm off to deal with EOS

**Comment by kevin-bates** [comment]
Thanks for the update @nifuki - please keep us posted on what you find.
